android/src/test/java/com/amplitude/android/plugins/AndroidLifecyclePluginTest.kt

```diff
@@ -0,0 +1,499 @@
+        }
+    }
+
+    @Test
+    fun `test application opened event is not tracked when disabled`() = runTest {
+        setDispatcher(testScheduler)
+        (amplitude.configuration as Configuration).trackingAppLifecycleEvents = false
+        amplitude.add(androidLifecyclePlugin)
+
+        val mockedPlugin = spyk(StubPlugin())
+        amplitude.add(mockedPlugin)
+        amplitude.isBuilt.await()
+
+        val mockedActivity = mockk<Activity>()
+        val mockedBundle = mockk<Bundle>()
+        androidLifecyclePlugin.onActivityCreated(mockedActivity, mockedBundle)
+        androidLifecyclePlugin.onActivityStarted(mockedActivity)
+        androidLifecyclePlugin.onActivityResumed(mockedActivity)
+
+        advanceUntilIdle()
+        Thread.sleep(100)
+
+        val tracks = mutableListOf<BaseEvent>()
+        verify(exactly = 0) { mockedPlugin.track(capture(tracks)) }
+        Assertions.assertEquals(0, tracks.count())
+    }
+
+    @Test
+    fun `test application backgrounded event is tracked`() = runTest {
+        setDispatcher(testScheduler)
+        (amplitude.configuration as Configuration).trackingAppLifecycleEvents = true
+        amplitude.add(androidLifecyclePlugin)
+
+        val mockedPlugin = spyk(StubPlugin())
+        amplitude.add(mockedPlugin)
+        amplitude.isBuilt.await()
+
+        val mockedActivity = mockk<Activity>()
+        androidLifecyclePlugin.onActivityPaused(mockedActivity)
+        androidLifecyclePlugin.onActivityStopped(mockedActivity)
+        androidLifecyclePlugin.onActivityDestroyed(mockedActivity)
+
+        advanceUntilIdle()
+        Thread.sleep(100)
+
+        val tracks = mutableListOf<BaseEvent>()
+        verify { mockedPlugin.track(capture(tracks)) }
+        Assertions.assertEquals(1, tracks.count())
+
+        with(tracks[0]) {
+            Assertions.assertEquals(DefaultEventUtils.EventTypes.APPLICATION_BACKGROUNDED, eventType)
+        }
+    }
+
+    @Test
+    fun `test application backgrounded event is not tracked when disabled`() = runTest {
+        setDispatcher(testScheduler)
+        (amplitude.configuration as Configuration).trackingAppLifecycleEvents = false
+        amplitude.add(androidLifecyclePlugin)
+
+        val mockedPlugin = spyk(StubPlugin())
+        amplitude.add(mockedPlugin)
+        amplitude.isBuilt.await()
+
+        val mockedActivity = mockk<Activity>()
+        androidLifecyclePlugin.onActivityPaused(mockedActivity)
+        androidLifecyclePlugin.onActivityStopped(mockedActivity)
+        androidLifecyclePlugin.onActivityDestroyed(mockedActivity)
+
+        advanceUntilIdle()
+        Thread.sleep(100)
+
+        val tracks = mutableListOf<BaseEvent>()
+        verify(exactly = 0) { mockedPlugin.track(capture(tracks)) }
+        Assertions.assertEquals(0, tracks.count())
+    }
+
+    @Test
+    fun `test screen viewed event is tracked`() = runTest {
```

Comment:
Nit: to make the tests more concise/maintainable maybe we could use `@TestFactory` to test enabled/disabled with the same logic vs multiple `@Test`s.

https://proandroiddev.com/kotlin-unit-tests-with-parameters-e37aab2b36f6

Reply:
Thanks for sharing this! I checked this article for example too: https://www.arhohuttunen.com/junit-5-kotlin/. I think it might not working in this case, as I want to check the detailed event type/properties when enabled, but not for disabled, and for each event, the flow (trigger) is different. Let me know if I miss any advanced usage of `TestFactory`.

Reply:
I was thinking you could switch on the provided parameter to determine the correct asserts for the results. Seems the initial setup and mocking for both the enabled/disabled test are the same, so would reduce a lot of setup code across the tests.

e.g.

```
(enabled: Bool) ->
// common setup
...
(amplitude.configuration as Configuration).trackingScreenViews = true
...
val tracks = mutableListOf<BaseEvent>()

// different results
if (enabled) { 
  verify { mockedPlugin.track(capture(tracks)) }
          Assertions.assertEquals(1, tracks.count())
          with(tracks[0]) {
              Assertions.assertEquals(DefaultEventUtils.EventTypes.SCREEN_VIEWED, eventType)
              Assertions.assertEquals(eventProperties?.get(DefaultEventUtils.EventProperties.SCREEN_NAME), "test-label")
} else {
    verify(exactly = 0) { mockedPlugin.track(capture(tracks)) }
    Assertions.assertEquals(0, tracks.count())
}
```

Reply:
Thanks @justin-fiedler for sharing this! I tried it, seems `dynamicTest` is not compatible with `runTest`, also I feel it actually worse the readability in this case, will keep the current tests then.